File: 0303_Kostebelova_EK/lab3/src/lockfree_queue.cpp
#include <atomic>
#include <memory>
#include <utility>
#include <chrono>

using namespace std;

template<typename T>
class lockfree_queue {
    struct Node {
        shared_ptr<T> data;
        atomic<Node*> next = nullptr;
    };

    atomic<Node*> head;
    atomic<Node*> tail;
public:
    lockfree_queue()
        : head(new Node()), tail(head.load()) {}

    void add(T element) {
        Node* new_node = new Node();
        new_node->data = make_shared<T>(move(element));
        while (true) {
            Node* last = tail.load();
            Node* next = last->next.load();
            if (last == tail.load()) {
                if (next == nullptr) {
                    if (last->next.compare_exchange_weak(next, new_node)) {
                        tail.compare_exchange_weak(last, new_node);
                        return;
                    }
                } else {
                    tail.compare_exchange_weak(last, next);
                }
            }

        }
    }

    bool get(T& element) {
        while (true) {
            Node* first = head.load();
            Node* last = tail.load();
            Node* next = first->next.load();
            if (first == last) {
                if (next == nullptr) {
                    return false;
                }
                tail.compare_exchange_weak(last, next);
            } else {
                if (head.compare_exchange_weak(first, next)) {
                    element = move(*next->data.get());
                    return true;
                }
            }
        }

    }

    ~lockfree_queue() {
        Node* cur = head.load();
        Node* tmp;
        while (cur) {
            tmp = cur;
            cur = cur->next;
            delete tmp;
        }

    }

};

File: 0303_Tabakov_PL_lab3/src/lab3/LockFreeQueue.java
package lab3;

import java.util.concurrent.atomic.AtomicReference;

public class LockFreeQueue<T> implements CustomQueue<T> {
    private static class Node<T> {
        private final AtomicReference<Node<T>> next = new AtomicReference<>();
        private T data;

        Node(T data) {
            this.data = data;
        }
    }

    private final AtomicReference<Node<T>> head;
    private final AtomicReference<Node<T>> tail;

    public LockFreeQueue() {
        Node<T> dummyNode = new Node<>(null);
        head = new AtomicReference<>(dummyNode);
        tail = new AtomicReference<>(dummyNode);
    }

    public void add(T element) {
        Node<T> new_node = new Node<>(element);
        while (true) {
            Node<T> last = tail.get();
            Node<T> next = last.next.get();
            if (last == tail.get()) {
                if (next == null) {
                    if (last.next.compareAndSet(next, new_node)) {
                        tail.compareAndSet(last, new_node);
                        return;
                    }
                } else {
                    tail.compareAndSet(last, next);
                }
            }
        }
    }

    public T take() {
        while (true) {
            Node<T> first = head.get();
            Node<T> last = tail.get();
            Node<T> next = first.next.get();
            if (first == last) {
                if (next == null) {
                    return null;
                }
                tail.compareAndSet(last, next);
            } else {
                if (head.compareAndSet(first, next)) {
                    return next.data;
                }
            }
        }
    }
}
